### PR TITLE
Repair{1} transform: remove unused vertices

### PR DESCRIPTION
**Pull request by stla**
As discussed in #354

### CODE REVIEW

File: src/Meshes.jl [hunk 1]
```diff
@@ -405,6 +405,7 @@ export
   LambdaMuSmoothing,
   LaplaceSmoothing,
   TaubinSmoothing,
+  Repair,
```
Comment (juliohm) [inline src/Meshes.jl]: Move to after StdCoords

File: src/transforms.jl [hunk 2]
```diff
@@ -98,3 +98,4 @@ include("transforms/translate.jl")
 include("transforms/stretch.jl")
 include("transforms/stdcoords.jl")
 include("transforms/smoothing.jl")
+include("transforms/repair.jl")
```
Comment (juliohm) [inline src/transforms.jl:100]: move to after stdcoords.jl

File: src/transforms/repair.jl [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+- K = 7: faces are coherently oriented
+"""
+struct Repair{K} <: StatelessGeometricTransform end
+
+# implement operation K = 1
+function apply(transform::Repair{1}, mesh)
+  # get the faces
+  topo = convert(HalfEdgeTopology, topology(mesh))
+  ∂₂₀ = Boundary{2,0}(topo)
+  nfaces = nelements(topo)
```
Comment (juliohm) [inline src/transforms/repair.jl]:
```suggestion
  nelem = nelements(topo)
```

File: src/transforms/repair.jl [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+  # indices of used vertices will be stored in this vector
+  used = Vector{Int}(undef, 0)
+
+  # initialize dictionary to map old indices to new indices
+  newindices = Dict{Int,Int}()
+
+  # to store the connectivities of the cleaned mesh
+  connectivities = Vector{Connectivity}(undef, 0)
```
Comment (juliohm) [inline src/transforms/repair.jl]:
```suggestion
  connec = Vector{Connectivity}(undef, 0)
```

File: src/transforms/repair.jl [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+  # l = length(used) will be incremented
+  l = 0
```
Comment (juliohm) [inline src/transforms/repair.jl]:
```suggestion
  nused = 0
```

File: src/transforms/repair.jl [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+  # iterate over faces
+  for f in 1:nfaces
```
Comment (juliohm) [inline src/transforms/repair.jl]:
```suggestion
  for e in 1:nelem
```

File: src/transforms/repair.jl [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    face = ∂₂₀(f)
+    for index in face
+      if !in(index, used)
```
Comment (juliohm) [inline src/transforms/repair.jl]: please use the latex \notin insted of !in

File: src/transforms/repair.jl [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+        l = l + 1
+        newindices[index] = l
+      end
+    end
+    connec = connect(tuple([newindices[i] for i in face]...))
+    push!(connectivities, connec)
+  end
+
+  # output mesh
+  SimpleMesh(vertices(mesh)[used], connectivities)
```
Comment (juliohm) [inline src/transforms/repair.jl]: create a new variable points = vertices(mesh)[used] before passing it to the SimpleMesh

File: test/transforms.jl [hunk 9]
```diff
@@ -81,4 +81,14 @@
...
+
+  @testset "Repair{1} (unused vertices)" begin
+    # a tetrahedron with an unused vertex
+    verts = P3.([(0, 0, 0), (0, 0, 1), (5, 5, 5), (0, 1, 0), (1, 0, 0)])
+    mesh = SimpleMesh(verts, connect.([(1, 2, 4), (1, 2, 5), (1, 4, 5), (2, 4, 5)], Triangle))
+    rmesh = mesh |> Repair{1}
+    @test nvertices(rmesh) == nvertices(mesh) - 1
+    @test nelements(rmesh) == nelements(mesh)
+    @test topology(rmesh) == topology(mesh)
+  end
```
Comment (juliohm) [inline test/transforms.jl]: move testset to before Smoothing testset

**Review by juliohm — CHANGES_REQUESTED** [review]
Nice addition. Review attached.

File: src/transforms/repair.jl (hunk 8 above)
Comment (juliohm) [reply src/transforms/repair.jl]: also, the apply function returns the result and the cache to revert the operation when that is possible. if that is not possible, we return `nothing`.

File: test/transforms.jl [hunk 10]
```diff
@@ -81,4 +81,14 @@
     @test nelements(smesh) == nelements(mesh)
     @test topology(smesh) == topology(mesh)
   end
+
+  @testset "Repair{1} (unused vertices)" begin
+    # a tetrahedron with an unused vertex
+    verts = P3.([(0, 0, 0), (0, 0, 1), (5, 5, 5), (0, 1, 0), (1, 0, 0)])
+    mesh = SimpleMesh(verts, connect.([(1, 2, 4), (1, 2, 5), (1, 4, 5), (2, 4, 5)], Triangle))
+    rmesh = mesh |> Repair{1}
```
Comment (juliohm) [inline test/transforms.jl]:
```suggestion
    rmesh = mesh |> Repair{1}()
```

File: test/transforms.jl [hunk 11]
```diff
@@ -61,7 +61,17 @@
     vnew2 = TB.reapply(trans, vset, cache)
     @test vnew == vnew2
   end
-  
+
+  @testset "Repair{1}" begin
+    # a tetrahedron with an unused vertex
+    verts = P3.([(0, 0, 0), (0, 0, 1), (5, 5, 5), (0, 1, 0), (1, 0, 0)])
+    mesh = SimpleMesh(verts, connect.([(1, 2, 4), (1, 2, 5), (1, 4, 5), (2, 4, 5)], Triangle))
```
Comment (juliohm) [inline test/transforms.jl]:
```suggestion
    mesh  = SimpleMesh(verts, connect.([(1, 2, 4), (1, 2, 5), (1, 4, 5), (2, 4, 5)], Triangle))
```

**Comment by stla** [comment]
Thanks. Looks like there's an error: `nvertices(rmesh)` is equal to 3. That's weird, I tested this function several times and I didn't notice that.

**Comment by stla** [comment]
That makes no sense :-( When I test the example outside the package there are 4 vertices, not 3.

File: src/transforms/repair.jl [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+  # indices of used vertices will be stored in this vector
+  used = Vector{Int}(undef, 0)
+
+  # initialize dictionary to map old indices to new indices
+  newindices = Dict{Int,Int}()
+
+  # to store the connectivities of the cleaned mesh
+  connectivities = Vector{Connectivity}(undef, 0)
+
+  # iterate over faces
```
Comment (juliohm) [inline src/transforms/repair.jl]:
```suggestion
  # iterate over elements
```

**Comment by stla** [comment]
Did you change something on my branch? My push has been rejected.

**Comment by juliohm** [comment]
Yes, I've been updating the PR here as you can see in the history above. Please try to pull the changes first before committing again or press . on GitHub so that it opens the editor in the PR version online.

**Comment by stla** [comment]
Ok. In fact `rmesh` is  not a mesh. `topology(rmesh)` fails as well as `nelements(rmesh)`.

What is `rmesh, nothing`?

**Comment by juliohm** [comment]
The apply function returns the result of the transform and a cache to
revert the transform. In this case the cache is nothing because the revert
is not implemented.

Em sex., 28 de abr. de 2023 11:49, stla ***@***.***> escreveu:

**Comment by stla** [comment]
Ok. It seems to work now. But `topology(rmesh) == topology(mesh)` is not true because the vertex indices have been adjusted. I'm going to push.

File: src/transforms/repair.jl [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+
+# implement operation K = 1
+function apply(transform::Repair{1}, mesh)
+  # get the faces
+  topo = convert(HalfEdgeTopology, topology(mesh))
+  ∂₂₀ = Boundary{2,0}(topo)
+  nelem = nelements(topo)
+
+  # indices of used vertices will be stored in this vector
+  used = Vector{Int}(undef, 0)
```
Comment (juliohm) [inline src/transforms/repair.jl]: replace this by a Set{Int} instead, which as fast checks for `in`

File: src/transforms/repair.jl [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+  for e in 1:nelem
+    elem = ∂₂₀(e)
+    for i in elem
+      if i ∉ used
+        push!(used, i)
+        nused += 1
+        newindices[i] = nused
+      end
+    end
+    c = connect(tuple([newindices[i] for i in elem]...))
```
Comment (juliohm) [inline src/transforms/repair.jl]: what happens when the dictionary doesn't have the index i?

Comment (juliohm) [reply src/transforms/repair.jl]: also, use ntuple(i->newindices[i], elem) instead to create the tuple statically

File: src/transforms/repair.jl [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+  # get the faces
+  topo = convert(HalfEdgeTopology, topology(mesh))
+  ∂₂₀ = Boundary{2,0}(topo)
+  nelem = nelements(topo)
+
+  # indices of used vertices will be stored in this vector
+  used = Vector{Int}(undef, 0)
+
+  # initialize dictionary to map old indices to new indices
+  newindices = Dict{Int,Int}()
```
Comment (juliohm) [inline src/transforms/repair.jl]: rename this variable to simply inds

File: test/transforms.jl [hunk 16]
```diff
@@ -61,7 +61,16 @@
     vnew2 = TB.reapply(trans, vset, cache)
     @test vnew == vnew2
   end
-  
+
+  @testset "Repair{1}" begin
+    # a tetrahedron with an unused vertex
+    verts = P3.([(0, 0, 0), (0, 0, 1), (5, 5, 5), (0, 1, 0), (1, 0, 0)])
```
Comment (juliohm) [inline test/transforms.jl]:
```suggestion
    verts = P3[(0, 0, 0), (0, 0, 1), (5, 5, 5), (0, 1, 0), (1, 0, 0)]
```

File: test/transforms.jl [hunk 17]
```diff
@@ -61,7 +61,16 @@
...
+    mesh  = SimpleMesh(verts, connect.([(1, 2, 4), (1, 2, 5), (1, 4, 5), (2, 4, 5)], Triangle))
```
Comment (juliohm) [inline test/transforms.jl]: move this connect argument to a variable connec before creating the mesh

**Review by juliohm — CHANGES_REQUESTED** [review]
A few more changes needed before we can merge.

**Comment by juliohm** [comment]
Also, please add the new Repair transform to the list of transforms in the docs after the changes above are made.

File: src/transforms/repair.jl (hunk 14 above)
Comment (stla) [reply src/transforms/repair.jl]: I believe it's not possible that the dictionary doesn't have the index i (it's late and I can't seriously think about that right now).

Your code `ntuple(i->newindices[i], elem)` does not work.

File: src/transforms/repair.jl (hunk 13 above)
Comment (stla) [reply src/transforms/repair.jl]: We cannot do `vertices(mesh)[used]` if `used` is a set.

**Review by juliohm — APPROVED** [review]

**Pull request merged**